heat/core/tests/test_manipulations.py

```diff
@@ -315,6 +316,273 @@ def test_concatenate(self):
+            )
+        )
+        res = ht.diagonal(a, offset=-1)
+        self.assertTrue(
+            torch.equal(
+                res._DNDarray__array,
+                torch.arange(size + 1).repeat(size).reshape(size, size + 1).t(),
+            )
+        )
+
+        res = ht.diagonal(a, dim1=1, dim2=2)
+        self.assertTrue(
+            torch.equal(
+                res._DNDarray__array,
+                torch.arange(size + 1).repeat(size + 1).reshape(size + 1, size + 1),
+            )
+        )
+        res = ht.diagonal(a, offset=1, dim1=1, dim2=2)
+        self.assertTrue(
+            torch.equal(
+                res._DNDarray__array,
+                torch.arange(1, size + 1).repeat(size + 1).reshape(size + 1, size),
+            )
+        )
+        res = ht.diagonal(a, offset=-1, dim1=1, dim2=2)
+        self.assertTrue(
+            torch.equal(
+                res._DNDarray__array, torch.arange(size).repeat(size + 1).reshape(size + 1, size)
+            )
+        )
+
+        res = ht.diagonal(a, dim1=0, dim2=2)
+        self.assertTrue(
+            torch.equal(
+                res._DNDarray__array,
+                torch.arange(size + 1).repeat(size + 1).reshape(size + 1, size + 1),
+            )
+        )
+
+        a = ht.array(data, split=0)
+        res = ht.diagonal(a, offset=1, dim1=0, dim2=1)
+        res.balance_()
+        self.assertTrue(
+            torch.equal(res._DNDarray__array, torch.arange(size + 1).reshape(size + 1, 1))
+        )
+        self.assertEqual(res.split, 1)
+
+        res = ht.diagonal(a, offset=-1, dim1=0, dim2=1)
+        res.balance_()
+        self.assertTrue(
+            torch.equal(res._DNDarray__array, torch.arange(size + 1).reshape(size + 1, 1))
+        )
+        self.assertEqual(res.split, 1)
+
+        res = ht.diagonal(a, offset=size + 1, dim1=0, dim2=1)
+        res.balance_()
+        self.assertTrue(
+            torch.equal(res._DNDarray__array, torch.empty((size + 1, 0), dtype=torch.int64))
+        )
+        self.assertTrue(res.shape[res.split] == 0)
+
+        with self.assertRaises(ValueError):
+            ht.diagonal(a, offset=None)
+
+        with self.assertRaises(ValueError):
+            ht.diagonal(a, dim1=1, dim2=1)
+
+        with self.assertRaises(ValueError):
+            ht.diagonal(a, dim1=1, dim2=-2)
+
+        with self.assertRaises(ValueError):
+            ht.diagonal(data)
+
+        self.assert_func_equal(
+            (5, 5, 5),
+            heat_func=ht.diagonal,
+            numpy_func=np.diagonal,
+            heat_args={"dim1": 0, "dim2": 2},
+            numpy_args={"axis1": 0, "axis2": 2},
+        )
+
+        self.assert_func_equal(
+            (5, 4, 3, 2),
+            heat_func=ht.diagonal,
+            numpy_func=np.diagonal,
+            heat_args={"dim1": 1, "dim2": 2},
+            numpy_args={"axis1": 1, "axis2": 2},
+        )
+
+        self.assert_func_equal(
```

Comment:
do these functions test multiple types? i dont remember. if so then everything is all good, but if not then please add a couple tests with different dtypes

Reply:
I added that, see #420

Reply:
I added that, see #420